.circleci/tests/test_despike.py

```diff
@@ -6,14 +6,121 @@
 Arguments have to be passed to these functions because the data may be
 mounted in a container somewhere unintuitively.
 """
-import os.path as op
+import os
+import numpy as np
+import nibabel as nb
+from nipype.pipeline import engine as pe
+from xcp_d.interfaces.regression import CiftiDespike
+from xcp_d.utils.write_save import read_ndata, write_ndata
+from xcp_d.utils.plot import _get_tr
 
 
-def test_data_availability(data_dir, working_dir, output_dir):
-    """Makes sure that we have access to all the testing data."""
-    assert op.exists(output_dir)
-    assert op.exists(working_dir)
-    assert op.exists(data_dir)
-    boldfile = data_dir + "/withoutfreesurfer/sub-01/func/" \
-        "sub-01_task-mixedgamblestask_run-1_space-MNI152NLin2009cAsym_desc-preproc_bold.nii.gz"
-    assert op.exists(boldfile)
+def test_nifti_despike(data_dir, tmp_path_factory):
+    """
+    Test Nifti despiking.
+
+    Confirm that the maximum and minimum voxel values decrease
+    after despiking.
+    """
+    # Read in the necessary inputs
+    boldfile = os.path.join(
+        data_dir, "withoutfreesurfer/sub-01/func/"
+        "sub-01_task-mixedgamblestask_run-1_space-"
+        "MNI152NLin2009cAsym_desc-preproc_bold.nii.gz"
+    )
+    maskfile = os.path.join(
+        data_dir, "withoutfreesurfer/sub-01/func/"
+        "sub-01_task-mixedgamblestask_run-1_space-"
+        "MNI152NLin2009cAsym_desc-brain_mask.nii.gz"
+    )
+    # Create some spikes in the second voxel
+    file_data = read_ndata(boldfile, maskfile)
+    voxel_data = file_data[2, :]  # a voxel across time
+    voxel_data_mean = np.mean(voxel_data)
+    voxel_data_std = np.std(voxel_data)
+    voxel_data[2] = voxel_data_mean + 10 * voxel_data_std
+    voxel_data[3] = voxel_data_mean - 10 * voxel_data_std
+    # What's the min and max, i.e: the amplitude of the spikes created?
+    spiked_min = min(voxel_data)
+    spiked_max = max(voxel_data)
+    # Let's write this temp file out for despiking
+    file_data[2, :] = voxel_data
+    tempdir = tmp_path_factory.mktemp("test_despike_nifti")
+    os.chdir(tempdir)
+    filename = "spikedfile.nii.gz"
+    write_ndata(
+        data_matrix=file_data,
+        mask=maskfile,
+        template=boldfile,
+        TR=0.8,
+        filename=filename,
+    )
+    spikedfile = filename
+    # Let's despike the image and write it out to a temp file
+    tempdir = os.getcwd()
+    os.system("3dDespike -NEW -prefix " + tempdir + "/3dDespike.nii.gz " + spikedfile)
```

Comment:
This is a version that works for me, using the Despike interface. I also think we should just drop the DespikePatch interface and use the nipype one directly.

```python
def test_nifti_despike(data_dir, tmp_path_factory):
    """Test Nifti despiking.

    Confirm that the maximum and minimum voxel values decrease
    after despiking.
    """
    # Read in the necessary inputs
    boldfile = os.path.join(
        data_dir,
        "withoutfreesurfer/sub-01/func",
        "sub-01_task-mixedgamblestask_run-1_space-MNI152NLin2009cAsym_desc-preproc_bold.nii.gz",
    )
    maskfile = os.path.join(
        data_dir,
        "withoutfreesurfer/sub-01/func",
        "sub-01_task-mixedgamblestask_run-1_space-MNI152NLin2009cAsym_desc-brain_mask.nii.gz",
    )

    # Create some spikes in the second voxel
    file_data = read_ndata(boldfile, maskfile)
    voxel_data = file_data[2, :]  # a voxel across time
    voxel_data_mean = np.mean(voxel_data)
    voxel_data_std = np.std(voxel_data)
    voxel_data[2] = voxel_data_mean + 10 * voxel_data_std
    voxel_data[3] = voxel_data_mean - 10 * voxel_data_std

    # What's the min and max, i.e: the amplitude of the spikes created?
    spiked_min = np.min(voxel_data)
    spiked_max = np.max(voxel_data)

    # Let's write this temp file out for despiking
    file_data[2, :] = voxel_data
    tempdir = tmp_path_factory.mktemp("test_despike_nifti")
    spikedfile = os.path.join(tempdir, "spikedfile.nii.gz")
    despiked_file = os.path.join(tempdir, "despikedfile.nii.gz")
    write_ndata(
        data_matrix=file_data,
        mask=maskfile,
        template=boldfile,
        TR=0.8,
        filename=spikedfile,
    )

    # Let's despike the image and write it out to a temp file
    despike_nifti = Despike(outputtype="NIFTI_GZ", args="-NEW")
    despike_nifti.inputs.in_file = spikedfile
    despike_nifti.inputs.out_file = despiked_file
    res = despike_nifti.run()

    assert os.path.isfile(despiked_file)

    file_data = read_ndata(res.outputs.out_file, maskfile)
    voxel_data = file_data[2, :]

    # What's the min and max of the despiked file?
    despiked_min = np.min(voxel_data)
    despiked_max = np.max(voxel_data)

    # Have the spikes been reduced?
    # i.e: has the minimum increased after despiking, and the maximum decreased?
    # Additionally, have the affines changed?
    assert spiked_min < despiked_min
    assert spiked_max > despiked_max

    # Are the affines the same before and after despiking?
    assert np.array_equal(nb.load(despiked_file).affine, nb.load(boldfile).affine)
```